crates/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -122,21 +146,115 @@ impl PerBlockContext {
+            }
+
+            // If no load values to substitute have been found in the current block, check the block's predecessors.
+            let predecessors = self.cfg.predecessors(block);
+            for predecessor in predecessors {
+                if dom_tree.is_reachable(predecessor)
+                    && dom_tree.dominates(predecessor, block)
+                    && !visited.contains(&predecessor)
+                {
+                    stack.push(predecessor);
+                }
+            }
+        }
+        address
+    }
+
+    // This method determines which loads should be substituted.
+    // Starting at the block supplied as a parameter, we check whether a store has occurred with the given address.
+    // If no store has occurred in the supplied block, the CFG is analyzed to determine whether
+    // a predecessor block has a store at the given address.
+    fn find_load_to_substitute(
+        &mut self,
+        block_id: BasicBlockId,
+        address: ValueId,
+        dfg: &DataFlowGraph,
+        instruction_id: &InstructionId,
+        loads_to_substitute: &mut BTreeMap<InstructionId, ValueId>,
+        load_values_to_substitute_per_block: &mut BTreeMap<ValueId, ValueId>,
+    ) -> bool {
+        let mut stack = vec![block_id];
+        let mut visited = HashSet::new();
+
+        let mut dom_tree = DominatorTree::with_cfg_and_post_order(&self.cfg, &self.post_order);
+        while let Some(block) = stack.pop() {
+            visited.insert(block);
+
+            for l in self.loops.yet_to_unroll.iter() {
+                // We do not want to substitute loads that take place within loops as this pass
+                // can occur before loop unrolling
+                if block == l.header {
+                    return false;
+                }
+            }
+
+            // Check whether there has been a store instruction in the current block
+            // If there has been a store, add a load to be substituted.
+            if let Some(last_value) = self.last_stores_with_block.get(&(address, block)) {
+                let result_value = *dfg
+                    .instruction_results(*instruction_id)
+                    .first()
+                    .expect("ICE: Load instructions should have single result");
+
+                loads_to_substitute.insert(*instruction_id, *last_value);
+                load_values_to_substitute_per_block.insert(result_value, *last_value);
+                self.load_values_to_substitute_per_func.insert(result_value, (*last_value, block));
+                return true;
+            }
+
+            // If no stores have been found in the current block, check the block's predecessors.
+            let predecessors = self.cfg.predecessors(block);
+            for predecessor in predecessors {
+                // TODO: Do I need is_reachable here? We are looping over only the reachable blocks but does
+                // that include a reachable block's predecessors?
```

Comment:
Wasn't sure if I needed is_reachable here

Reply:
It could theoretically be necessary if there were previous blocks that were removed but still jump to any of these blocks internally. I'm not sure if this ever occurs on any of our optimization passes but it's safer to have the check I suppose.

Can `PostOrder` be used at all since we're iterating backwards? Or not because we have to start on a given block id?

Reply:
I am kind of using `PostOrder` by using the dom tree no?